yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+
+```
+kernel_public_inputs.gas_limits := private_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used :=
+    FIXED_DA_GAS
+    + private_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.compute_gas_used :=
+    private_kernel_public_inputs.end.compute_gas_used
+
+# make assertions that gas limits are not exceeded
+
+# purely private transactions cannot pay any fee
+kernel_public_inputs.transaction_fee := 0
+```
+
+# Transaction with public component
+
+If a transaction has a public component, `end_non_revertible.gas_used.da_gas_used` is initialized to `FIXED_DA_GAS` in the `PrivateKernelTailToPublic` kernel.
+
+### Interaction with the public VM
+
+The public VM retains the current amount of `daGasLeft` in its [MachineState](../public-vm/state.md#machine-state).
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+```
+
+At the end of execution, the public VM will have an updated `daGasLeft` value.
+
+This value will be used by the public kernel to update the `da_gas_used` in the `PublicAccumulatedData` struct corresponding to the current phase:
+
+```
+# if in setup 
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end.gas_used.da_gas_used
+
+# if in app logic
+public_kernel_public_inputs.end.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+```
+
+
+It will be the same for the `compute_gas_used`.
+
+### Out of gas
+
+Public execution must stop as soon as possible once `daGasLeft` is exhausted.
+
+The public VM will revert if an exception is thrown, including an out of gas exception.
+When this occurs, the public VM will return a revert code.
+
+If this occurs in setup, the transaction is invalid and cannot be included in the block.
+
+If this occurs in app_logic, `daGasLeft` is set to 0, and thus `public_kernel_public_inputs.gas_limits.da_gas_limit == public_kernel_public_inputs.end.gas_used.da_gas_used + public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used`. 
+
+Similarly for `computeGasLeft`.
+
+In the teardown phase, there cannot be an out of gas error because the teardown phase effectively has infinite gas (since the user has prepaid for it via `inclusion_fee`). If there is any other exception though, the transaction is invalid and cannot be included in the block. It is thus expected that sequencers will maintain a whitelist of teardown functions and the minimum `inclusion_fee` required for each.
```

Comment:
The "infinite gas" part feels risky. This means that any full node that wants to follow along the chain by executing the L2 txs they see may be forced to execute an arbitrary amount of code, set at the discretion of a sequencer that decided to run it.

Reply:
Also, if we want to have a flat fee for the teardown phase (which sounds reasonable), it seems that should be a flat gas _limit_, since the fee per gas could still be variable, right?